Автоматическое многопоточное удаление expired событий по расписанию

vzeten opened the pull request:
AsyncConfig
✅ Создал `@Configuration` с `@EnableScheduling`
✅ Bean `ExecutorService` (размер pool из `application.yaml`)

EventScheduler
✅ `@Component` с методом `@Scheduled`
✅ Вызывает `EventService.removePastEvents()` по cron из конфига

EventService
✅ Находит expired события через `findByEventDateBefore()` ✅ Параллельное удаление батчей через `CompletableFuture` ✅ `join()` для ожидания завершения всех потоков

EventRepository
✅ Добавил `findByEventDateBefore(LocalDateTime)` для поиска expired событий

application.yaml
✅ `scheduler.thread-pool-size: 10`
✅ `scheduler.batch-size: 100`
✅ `scheduler.expired-events.cron: "0 0 2 * * *"`

***

**Реализовано**: Автоматическое многопоточное удаление expired событий по расписанию.

Commit:
AsyncConfig
✅ Создал `@Configuration` с `@EnableScheduling`
✅ Bean `ExecutorService` (размер pool из `application.yaml`)

EventScheduler
✅ `@Component` с методом `@Scheduled`
✅ Вызывает `EventService.removePastEvents()` по cron из конфига

EventService
✅ Находит expired события через `findByEventDateBefore()`
✅ Параллельное удаление батчей через `CompletableFuture`
✅ `join()` для ожидания завершения всех потоков

EventRepository
✅ Добавил `findByEventDateBefore(LocalDateTime)` для поиска expired событий

application.yaml
✅ `scheduler.thread-pool-size: 10`
✅ `scheduler.batch-size: 100`
✅ `scheduler.expired-events.cron: "0 0 2 * * *"`

***

**Реализовано**: Автоматическое многопоточное удаление expired событий по расписанию.

Commit:
AsyncConfig
✅ Создал `@Configuration` с `@EnableScheduling`
✅ Bean `ExecutorService` (размер pool из `application.yaml`)

EventScheduler
✅ `@Component` с методом `@Scheduled`
✅ Вызывает `EventService.removePastEvents()` по cron из конфига

EventService
✅ Находит expired события через `findByEventDateBefore()`
✅ Параллельное удаление батчей через `CompletableFuture`
✅ `join()` для ожидания завершения всех потоков

EventRepository
✅ Добавил `findByEventDateBefore(LocalDateTime)` для поиска expired событий

application.yaml
✅ `scheduler.thread-pool-size: 10`
✅ `scheduler.batch-size: 100`
✅ `scheduler.expired-events.cron: "0 0 2 * * *"`
EventServiceTest
✅ Добавил тесты
***

**Реализовано**: Автоматическое многопоточное удаление expired событий по расписанию.

Commit:
AsyncConfig
✅ Создал `@Configuration` с `@EnableScheduling`
✅ Bean `ExecutorService` (размер pool из `application.yaml`)

EventScheduler
✅ `@Component` с методом `@Scheduled`
✅ Вызывает `EventService.removePastEvents()` по cron из конфига

EventService
✅ Находит expired события через `findByEventDateBefore()`
✅ Параллельное удаление батчей через `CompletableFuture`
✅ `join()` для ожидания завершения всех потоков

EventRepository
✅ Добавил `findByEventDateBefore(LocalDateTime)` для поиска expired событий

application.yaml
✅ `scheduler.thread-pool-size: 10`
✅ `scheduler.batch-size: 100`
✅ `scheduler.expired-events.cron: "0 0 2 * * *"`
EventServiceTest
✅ Добавил тесты
***

**Реализовано**: Автоматическое многопоточное удаление expired событий по расписанию.

Kalex-k commented on src/main/java/school/faang/user_service/scheduler/Scheduler.java:
```diff
@@ -0,0 +1,17 @@
+public class Scheduler {
+    private final EventService eventService;
+
+    @Scheduled(cron = "${scheduler.cron}")
```
Не уверена, что путь указан правильно, лучше пиши полный путь scheduler.expired-events.cron

vzeten replied:
Спасибо! так и есть

Kalex-k commented on src/main/java/school/faang/user_service/service/event/EventService.java:
```diff
@@ -0,0 +1,61 @@
+    @Value("${scheduler.batch-size}")
+    private int batchSize;
+
+    public void removePastEvents() {
```
Отсутствует транзакция, может привести к потере целостности данных.

vzeten replied:
Тут специально, транзацию не ставил, если не удалит сейчас что-то из базы, удалит в следующий раз когда запустится

Kalex-k commented on src/main/java/school/faang/user_service/service/event/EventService.java:
```diff
@@ -0,0 +1,61 @@
+
+    public void removePastEvents() {
+        List<Event> expiredEvents = eventRepository.findByEndDateBefore(LocalDateTime.now());
+
```
я бы в репозиторий написfла кастомный метод, который будет получать только ID, это во много раз меньше будет потреблять память и быстрее передавать данные. Сейчас у тебя метод загружает полные Entity (неэффективно для массового удаления)

Kalex-k commented on src/main/java/school/faang/user_service/service/event/EventService.java:
```diff
@@ -0,0 +1,61 @@
+                        () -> eventRepository.deleteAllById(batch),
+                        executorService
+                ))
+                .toList();
```
Добавь обработку ошибок. Если внутри лямбды произойдет исключение во время асинхронного выполнения, то оно просто застрянет в Future.

Kalex-k commented on src/main/java/school/faang/user_service/service/event/EventService.java:
```diff
@@ -0,0 +1,61 @@
+                .toList();
+
+        CompletableFuture.allOf(futures.toArray(new CompletableFuture[0])).join();
+
```
Тоже нужен try-catch что бы при исключении увидеть какой батч упал

Commit:
Поправил комментарии

AlexeyAntipin commented on src/main/java/school/faang/user_service/config/AsyncConfig.java:
```diff
@@ -0,0 +1,18 @@
+
+    @Bean(destroyMethod = "shutdown")
+    public ExecutorService executorService(@Value("${scheduler.thread-pool-size}") int threadPoolSize) {
+        System.out.println("Spring создаёт ExecutorService");
```
В реальных приложениях вывод через System.out.println делать не стоит (зачастую даже на проектах это запрещено чекстайлом). Так ты не сможешь уровень логирования использовать + не сможешь писать логи асинхронно

Используй лучше log

AlexeyAntipin reviewed (commented):
Ты проверял, что шедулинг работает? Не вижу у тебя нигде аннотации @EnableScheduling на конфигурациях

Commit:
Добавил @EnableScheduling
и логирование в AsyncConfig

Commit:
Merge branch 'hydra-master-stream12' into hydra12-BJS2-95705-remove-past-events

Commit:
В этих тестах добавил аннотацию @Disabled :
PremiumServiceEdgeCasesTest.java
PremiumServiceIntegrationTest.java
PremiumServiceTest.java

Commit:
В этих тестах добавил аннотацию @Disabled :
PremiumServiceConcurrencyTest.java
PremiumServiceEdgeCasesTest.java
PremiumServiceIntegrationTest.java
PremiumServiceTest.java

vzeten committed:
Merge branch 'hydra-master-stream12' into hydra12-BJS2-95705-remove-past-events

mysosnovskaya commented on src/main/java/school/faang/user_service/service/event/EventService.java:
```diff
@@ -0,0 +1,72 @@
+        }
+    }
+
+    private List<List<Long>> partitionIntoBatches(List<Long> ids, int batchSize) {
+        List<List<Long>> batches = new ArrayList<>();
+        for (int i = 0; i < ids.size(); i += batchSize) {
+            batches.add(ids.subList(i, Math.min(i + batchSize, ids.size())));
+        }
+        return batches;
+    }
```
Можно использовать ListUtils.partition от apache, чтобы руками не разбивать.

mysosnovskaya reviewed (approved):
Написала коммент на будущее, в остальном ок
